File: nostr-knowledge/examples/basics/connect-relay/connect_relay.go
/*
NOSTR Connect to Relay — Go

Connects to a NOSTR relay via WebSocket, sends a REQ subscription,
receives events, and handles the EOSE (End Of Stored Events) message.

Dependencies:
	go get github.com/nbd-wtf/go-nostr

Run:
	go run connect_relay.go

References:
	- NIP-01: https://github.com/nostr-protocol/nips/blob/master/01.md
*/

package main

import (
	"context"
	"fmt"
	"log"
	"time"

	"github.com/nbd-wtf/go-nostr"
)

const relayURL = "wss://relay.damus.io"

func main() {
	fmt.Println("=== NOSTR Connect to Relay ===")
	fmt.Println()

	// Step 1: Connect to the relay.
	// This opens a WebSocket connection and performs the initial handshake.
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	fmt.Printf("Connecting to %s...\n", relayURL)
	relay, err := nostr.RelayConnect(ctx, relayURL)
	if err != nil {
		log.Fatalf("Failed to connect: %v", err)
	}
	defer relay.Close()
	fmt.Printf("Connected to %s\n\n", relay.URL)

	// Step 2: Define a subscription filter.
	// Filters tell the relay which events we want to receive.
	// This filter requests the 5 most recent kind 1 (text note) events.
	filters := nostr.Filters{
		{
			Kinds: []int{1}, // Kind 1 = text notes
			Limit: 5,        // Only return the 5 most recent
		},
	}

	// Step 3: Subscribe to events matching the filter.
	// The relay sends stored events first, then an EOSE signal,
	// then continues streaming new events in real-time.
	fmt.Printf("Subscribing with filter: kinds=[1], limit=5\n")
	fmt.Println("Waiting for events...\n")

	sub, err := relay.Subscribe(ctx, filters)
	if err != nil {
		log.Fatalf("Failed to subscribe: %v", err)
	}
	defer sub.Unsub()

	// Step 4: Receive events from the subscription.
	// The subscription has two channels:
	//   sub.Events — receives matching events
	//   sub.EndOfStoredEvents — signals that all stored events have been sent
	for {
		select {
		case event := <-sub.Events:
			// An event matching our filter.
			ts := time.Unix(int64(event.CreatedAt), 0).UTC()
			content := event.Content
			if len(content) > 100 {
				content = content[:100] + "..."
			}

			fmt.Println("--- Received Event ---")
			fmt.Printf("  ID:      %s\n", event.ID)
			fmt.Printf("  Author:  %s\n", event.PubKey)
			fmt.Printf("  Kind:    %d\n", event.Kind)
			fmt.Printf("  Time:    %s\n", ts.Format(time.RFC3339))
			fmt.Printf("  Content: %s\n", content)
			fmt.Println()

		case <-sub.EndOfStoredEvents:
			// All stored events have been sent.
			fmt.Println("=== EOSE (End of Stored Events) ===")
			fmt.Println("All stored events received. New events would stream in real-time.")
			fmt.Println("Closing subscription and disconnecting...")
			return

		case <-ctx.Done():
			fmt.Println("Context timeout reached.")
			return
		}
	}
}

File: nostr-knowledge/examples/basics/connect-relay/connect_relay.ts
/**
 * NOSTR Connect to Relay — TypeScript
 *
 * Connects to a NOSTR relay via WebSocket, sends a REQ subscription,
 * receives events, and handles the EOSE (End Of Stored Events) message.
 *
 * Dependencies:
 *   npm install nostr-tools ws
 *   npm install -D @types/ws
 *
 * Run:
 *   npx ts-node connect_relay.ts
 *
 * References:
 *   - NIP-01: https://github.com/nostr-protocol/nips/blob/master/01.md
 */

import { Relay } from "nostr-tools/relay";
import type { Filter } from "nostr-tools/filter";
import type { Event } from "nostr-tools/pure";

// For Node.js, we need a WebSocket polyfill.
// In browser environments, the native WebSocket is used automatically.
import WebSocket from "ws";
(globalThis as any).WebSocket = WebSocket;

const RELAY_URL = "wss://relay.damus.io";

async function main(): Promise<void> {
  console.log("=== NOSTR Connect to Relay ===\n");

  // Step 1: Connect to the relay.
  // This opens a WebSocket connection and waits for the connection to be established.
  console.log(`Connecting to ${RELAY_URL}...`);
  const relay = await Relay.connect(RELAY_URL);
  console.log(`Connected to ${relay.url}\n`);

  // Step 2: Define a subscription filter.
  // Filters tell the relay which events we want to receive.
  // This filter requests the 5 most recent kind 1 (text note) events.
  const filter: Filter = {
    kinds: [1],    // Kind 1 = text notes
    limit: 5,     // Only return the 5 most recent
  };

  // Step 3: Subscribe to events matching the filter.
  // The relay sends matching stored events, then an EOSE message,
  // then continues streaming new events that match in real-time.
  console.log("Subscribing with filter:", JSON.stringify(filter));
  console.log("Waiting for events...\n");

  const sub = relay.subscribe([filter], {
    // Called for each event received (both stored and real-time).
    onevent(event: Event) {
      console.log("--- Received Event ---");
      console.log(`  ID:      ${event.id}`);
      console.log(`  Author:  ${event.pubkey}`);
      console.log(`  Kind:    ${event.kind}`);
      console.log(`  Time:    ${new Date(event.created_at * 1000).toISOString()}`);
      console.log(`  Content: ${event.content.substring(0, 100)}${event.content.length > 100 ? "..." : ""}`);
      console.log();
    },

    // Called when the relay has sent all stored events matching the filter.
    // After EOSE, any new events are real-time.
    oneose() {
      console.log("=== EOSE (End of Stored Events) ===");
      console.log("All stored events received. New events would stream in real-time.");
      console.log("Closing subscription and disconnecting...\n");

      // Step 4: Clean up.
      sub.close();       // Unsubscribe (sends CLOSE to relay)
      relay.close();     // Close WebSocket connection
    },
  });
}

main().catch(console.error);
